File: apps/mobile/ios/Aurelia/ViewModels/HomeViewModel.swift
import AureliaCore
import Foundation
import Observation
import os

@Observable
final class HomeViewModel: @unchecked Sendable {
    var isLoading = false
    var error: String?
    var featuredAlbums: [FeaturedAlbum] = []
    var currentFeaturedIndex = 0
    var mostPlayed: [Song] = []
    var recentlyPlayed: [Song] = []
    var recentlyAddedAlbums: [AlbumItem] = []
    var randomAlbums: [AlbumItem] = []
    var nowPlaying: NowPlayingState?
    var currentSongId: String?

    private let sessionStore = SessionStore.shared
    private let logger = Logger(subsystem: "com.aurelia.app", category: "HomeViewModel")
    private var hasLoadedInitialData = false
    private var allSongs: [Song] = []
    private var songIdByTitleArtist: [String: String] = [:]
    private nonisolated static let useSharedHomeDerivation = true

    private struct HomeSections {
        let mostPlayed: [Song]
        let recentlyPlayed: [Song]
        let recentlyAddedAlbums: [AlbumItem]
        let randomAlbums: [AlbumItem]
        let featuredAlbums: [FeaturedAlbum]
    }

    private struct HomeSectionLimits: Sendable {
        let mostPlayed: Int
        let recentlyPlayed: Int
        let albumSection: Int
        let featuredAlbums: Int
    }

    func loadHomeData() {
        guard !hasLoadedInitialData else { return }

        isLoading = true
        error = nil

        let sectionLimits = HomeSectionLimits(
            mostPlayed: UIConstants.mostPlayedLimit,
            recentlyPlayed: UIConstants.recentlyPlayedLimit,
            albumSection: UIConstants.albumSectionLimit,
            featuredAlbums: UIConstants.featuredAlbumsLimit
        )

        Task.detached { [self] in
            guard let creds = await sessionStore.getCredentialsAsync(),
                  !creds.serverUrl.isEmpty, !creds.token.isEmpty, !creds.userId.isEmpty
            else {
                await MainActor.run {
                    self.isLoading = false
                    self.error = "Missing session data"
                }
                return
            }

            let appDataDir = await MainActor.run { sessionStore.getAppDataDir() ?? "" }
            let shouldRefresh = await sessionStore.shouldRefreshLibraryAsync()

            // Load cached data first
            var loadedCache = false
            if !appDataDir.isEmpty {
                do {
                    let cached = try loadCachedSongs(appDataDir: appDataDir)
                    if !cached.isEmpty {
                        loadedCache = true
                        let sections = Self.computeHomeSections(cached, limits: sectionLimits)
                        let songIdCache = Self.buildSongIdCache(cached)
                        await MainActor.run {
                            self.allSongs = cached
                            self.songIdByTitleArtist = songIdCache
                            self.applyHomeSections(sections)
                        }
                    }
                } catch {
                    logger.warning("Failed to load cached songs: \(error)")
                }
            }

            if loadedCache, !shouldRefresh {
                await MainActor.run {
                    self.hasLoadedInitialData = true
                }
                return
            }

            // Fetch fresh data
            do {
                let songs = try await fetchSongs(
                    serverUrl: creds.serverUrl,
                    token: creds.token,
                    userId: creds.userId,
                    appDataDir: appDataDir
                )
                let sections = Self.computeHomeSections(songs, limits: sectionLimits)
                let songIdCache = Self.buildSongIdCache(songs)
                await MainActor.run {
                    self.allSongs = songs
                    self.songIdByTitleArtist = songIdCache
                    self.applyHomeSections(sections)
                    self.hasLoadedInitialData = true
                    self.sessionStore.markLibraryRefreshed()
                }
            } catch {
                if await !AuthInterceptor.shared.handlePotentialAuthError(error) {
                    await MainActor.run {
                        self.isLoading = false
                        self.error = error.localizedDescription
                    }
                }
            }
        }
    }

    @MainActor
    private func applyHomeSections(_ sections: HomeSections) {
        mostPlayed = sections.mostPlayed
        recentlyPlayed = sections.recentlyPlayed
        recentlyAddedAlbums = sections.recentlyAddedAlbums
        randomAlbums = sections.randomAlbums
        featuredAlbums = sections.featuredAlbums
        isLoading = false
    }

    private nonisolated static func computeHomeSections(_ songs: [Song], limits: HomeSectionLimits) -> HomeSections {
        if useSharedHomeDerivation {
            let derived = deriveMobileHomeData(
                songs: songs,
                mostPlayedLimit: Int64(limits.mostPlayed),
                recentlyPlayedLimit: Int64(limits.recentlyPlayed),
                albumSectionLimit: Int64(limits.albumSection),
                featuredAlbumsLimit: Int64(limits.featuredAlbums)
            )

            return HomeSections(
                mostPlayed: derived.mostPlayed,
                recentlyPlayed: derived.recentlyPlayed,
                recentlyAddedAlbums: derived.recentlyAdded.map {
                    AlbumItem(
                        id: $0.id ?? "",
                        name: $0.name,
                        artist: $0.artist,
                        albumArtUrl: $0.albumArtUrl,
                        songCount: Int($0.songCount)
                    )
                },
                randomAlbums: derived.randomAlbums.map {
                    AlbumItem(
                        id: $0.id ?? "",
                        name: $0.name,
                        artist: $0.artist,
                        albumArtUrl: $0.albumArtUrl,
                        songCount: Int($0.songCount)
                    )
                },
                featuredAlbums: derived.featuredAlbums.map {
                    FeaturedAlbum(
                        id: $0.id ?? "",
                        name: $0.name,
                        artist: $0.artist,
                        albumArtUrl: $0.albumArtUrl,
                        songCount: Int($0.songCount)
                    )
                }
            )
        }

        // Most played
        let mostPlayed = songs
            .filter { ($0.playCount ?? 0) > 0 }
            .sorted { ($0.playCount ?? 0) > ($1.playCount ?? 0) }
            .prefix(limits.mostPlayed)
            .map(\.self)

        // Recently played
        let recentlyPlayed = songs
            .filter { $0.datePlayed?.isEmpty == false }
            .sorted { ($0.datePlayed ?? "") > ($1.datePlayed ?? "") }
            .prefix(limits.recentlyPlayed)
            .map(\.self)

        // Group by album
        let songsByAlbumId = songs.compactMap { song -> (String, Song)? in
            guard let albumId = song.albumId, !albumId.isEmpty else { return nil }
            return (albumId, song)
        }
        let albumsMap = Dictionary(grouping: songsByAlbumId, by: { $0.0 })

        // Recently added albums
        let recentlyAddedAlbums = albumsMap
            .map { albumId, albumEntries -> (AlbumItem, String) in
                let albumSongs = albumEntries.map(\.1)
                let firstSong = albumSongs.max(by: { ($0.dateCreated ?? "") < ($1.dateCreated ?? "") }) ?? albumSongs[0]
                return (
                    AlbumItem(
                        id: albumId,
                        name: firstSong.album ?? "Unknown Album",
                        artist: firstSong.artists?.first ?? "Unknown Artist",
                        albumArtUrl: firstSong.albumArtUrl,
                        songCount: albumSongs.count
                    ),
                    firstSong.dateCreated ?? ""
                )
            }
            .sorted { $0.1 > $1.1 }
            .prefix(limits.albumSection)
            .map(\.0)

        // Random albums
        let randomAlbums = albumsMap
            .map { albumId, albumEntries -> AlbumItem in
                let albumSongs = albumEntries.map(\.1)
                let firstSong = albumSongs[0]
                return AlbumItem(
                    id: albumId,
                    name: firstSong.album ?? "Unknown Album",
                    artist: firstSong.artists?.first ?? "Unknown Artist",
                    albumArtUrl: firstSong.albumArtUrl,
                    songCount: albumSongs.count
                )
            }
            .shuffled()
            .prefix(limits.albumSection)
            .map(\.self)

        // Featured albums
        let featuredAlbums = albumsMap
            .filter { $0.value.contains { $0.1.albumArtUrl?.isEmpty == false } }
            .map { albumId, albumEntries -> FeaturedAlbum in
                let albumSongs = albumEntries.map(\.1)
                let firstSong = albumSongs[0]
                return FeaturedAlbum(
                    id: albumId,
                    name: firstSong.album ?? "Unknown Album",
                    artist: firstSong.artists?.joined(separator: ", ") ?? "Unknown Artist",
                    albumArtUrl: firstSong.albumArtUrl,
                    songCount: albumSongs.count
                )
            }
            .shuffled()
            .prefix(limits.featuredAlbums)
            .map(\.self)

        return HomeSections(
            mostPlayed: mostPlayed,
            recentlyPlayed: recentlyPlayed,
            recentlyAddedAlbums: recentlyAddedAlbums,
            randomAlbums: randomAlbums,
            featuredAlbums: featuredAlbums
        )
    }

    // MARK: - Playback

    func playSongFromList(_ songId: String, songList: [Song], playerController: AudioPlayerController) {
        guard let serverUrl = sessionStore.serverUrl, let token = sessionStore.token else { return }
        guard let startIndex = songList.firstIndex(where: { $0.id == songId }) else { return }
        currentSongId = songId
        playerController.setQueue(songList, serverUrl: serverUrl, token: token, startIndex: startIndex)
    }

    func playAlbum(_ albumId: String, playerController: AudioPlayerController) {
        guard let serverUrl = sessionStore.serverUrl, let token = sessionStore.token else { return }
        let albumSongs = allSongs
            .filter { $0.albumId == albumId }
            .sorted { ($0.trackNumber ?? 0) < ($1.trackNumber ?? 0) }
        guard !albumSongs.isEmpty else { return }
        currentSongId = albumSongs[0].id
        playerController.setQueue(albumSongs, serverUrl: serverUrl, token: token)
    }

    func shuffleAlbum(_ albumId: String, playerController: AudioPlayerController) {
        guard let serverUrl = sessionStore.serverUrl, let token = sessionStore.token else { return }
        let albumSongs = allSongs.filter { $0.albumId == albumId }.shuffled()
        guard !albumSongs.isEmpty else { return }
        currentSongId = albumSongs[0].id
        playerController.setQueue(albumSongs, serverUrl: serverUrl, token: token)
    }

    // MARK: - Featured Navigation

    func nextFeaturedAlbum() {
        guard !featuredAlbums.isEmpty else { return }
        currentFeaturedIndex = (currentFeaturedIndex + 1) % featuredAlbums.count
    }

    func previousFeaturedAlbum() {
        guard !featuredAlbums.isEmpty else { return }
        currentFeaturedIndex = currentFeaturedIndex > 0 ? currentFeaturedIndex - 1 : featuredAlbums.count - 1
    }

    // MARK: - Helpers

    private nonisolated static func buildSongIdCache(_ songs: [Song]) -> [String: String] {
        var cache: [String: String] = [:]
        for song in songs {
            let key = "\(song.name)_\(song.artists?.first ?? "")"
            cache[key] = song.id
        }
        return cache
    }
}

File: apps/mobile/android/app/src/main/java/com/aurelia/app/ui/HomeViewModel.kt
package com.aurelia.app.ui

import android.os.SystemClock
import android.util.Log
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.aurelia.app.auth.AuthInterceptor
import com.aurelia.app.player.PlayerController
import com.aurelia.app.storage.SessionStore
import com.aurelia.app.utils.buildSongIdCache
import com.aurelia.app.utils.validateSession
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import uniffi.aurelia_core.AppException
import uniffi.aurelia_core.Song
import uniffi.aurelia_core.deriveMobileHomeData
import uniffi.aurelia_core.fetchSongs
import uniffi.aurelia_core.loadCachedSongs

class HomeViewModel(
  private val sessionStore: SessionStore,
  private val playerController: PlayerController,
) : ViewModel() {
  private val mutableState = MutableStateFlow(HomeState())
  val state: StateFlow<HomeState> = mutableState

  // All songs cache for queue building
  private var allSongs: List<Song> = emptyList()
  private val useSharedHomeDerivation = true

  // Cache for song ID lookup
  private var songIdByTitleArtist: Map<Pair<String, String>, String> = emptyMap()
  private var loadJob: Job? = null
  private var lastLoadedAtMs: Long = 0L

  private val nowPlayingMapper = NowPlayingMapper()

  init {
    viewModelScope.launch {
      playerController.snapshots.collect { snapshot ->
        if (!nowPlayingMapper.shouldUpdate(snapshot)) return@collect
        val (nowPlaying, songId) = nowPlayingMapper.mapToNowPlaying(snapshot, songIdByTitleArtist)
        mutableState.update { it.copy(nowPlaying = nowPlaying, currentSongId = songId) }
      }
    }
  }

  fun ensureLoaded(force: Boolean = false) {
    if (!force && loadJob?.isActive == true) return
    val hasData =
      mutableState.value.featuredAlbums.isNotEmpty() ||
        mutableState.value.mostPlayed.isNotEmpty() ||
        mutableState.value.recentlyPlayed.isNotEmpty()
    val isFresh = SystemClock.elapsedRealtime() - lastLoadedAtMs < LOAD_FRESHNESS_MS
    if (!force && hasData && isFresh) return

    val session = validateSession(sessionStore)
    if (session == null) {
      mutableState.update { it.copy(error = "Missing session data") }
      return
    }

    mutableState.update { it.copy(isLoading = true, error = null) }

    loadJob = viewModelScope.launch(Dispatchers.IO) {
      // Try loading from cache first
      if (!session.appDataDir.isNullOrBlank()) {
        try {
          val cachedSongs = loadCachedSongs(session.appDataDir)
          if (cachedSongs.isNotEmpty()) {
            allSongs = cachedSongs
            songIdByTitleArtist = buildSongIdCache(cachedSongs)
            processHomeData(cachedSongs)
          }
        } catch (e: Exception) {
          Log.w("HomeViewModel", "Failed to load cached songs", e)
        }
      }

      // Fetch fresh data
      try {
        val songs = fetchSongs(session.serverUrl, session.token, session.userId, session.appDataDir ?: "")
        allSongs = songs
        songIdByTitleArtist = buildSongIdCache(songs)
        processHomeData(songs)
        lastLoadedAtMs = SystemClock.elapsedRealtime()
      } catch (error: AppException) {
        if (!AuthInterceptor.handlePotentialAuthError(error.message)) {
          mutableState.update { it.copy(isLoading = false, error = error.message ?: "Failed to load") }
        }
      } catch (error: Exception) {
        if (!AuthInterceptor.handlePotentialAuthError(error)) {
          mutableState.update { it.copy(isLoading = false, error = "Failed to load") }
        }
      }
    }
  }

  fun loadHomeData() {
    ensureLoaded(force = false)
  }

  private fun processHomeData(songs: List<Song>) {
    if (useSharedHomeDerivation) {
      processHomeDataWithSharedDerivation(songs)
      return
    }

    processHomeDataLegacy(songs)
  }

  private fun processHomeDataWithSharedDerivation(songs: List<Song>) {
    val derived =
      deriveMobileHomeData(
        songs = songs,
        mostPlayedLimit = UiConstants.MOST_PLAYED_LIMIT.toLong(),
        recentlyPlayedLimit = UiConstants.RECENTLY_PLAYED_LIMIT.toLong(),
        albumSectionLimit = UiConstants.ALBUM_SECTION_LIMIT.toLong(),
        featuredAlbumsLimit = UiConstants.FEATURED_ALBUMS_LIMIT.toLong(),
      )

    mutableState.update {
      it.copy(
        isLoading = false,
        featuredAlbums = derived.featuredAlbums.map(::toFeaturedAlbum),
        mostPlayed = derived.mostPlayed,
        recentlyPlayed = derived.recentlyPlayed,
        recentlyAddedAlbums = derived.recentlyAdded.map(::toAlbumItem),
        randomAlbums = derived.randomAlbums.map(::toAlbumItem),
      )
    }
  }

  private fun processHomeDataLegacy(songs: List<Song>) {
    // Most played - top 10 by playCount
    val mostPlayed =
      songs
        .filter { (it.playCount ?: 0) > 0 }
        .sortedByDescending { it.playCount ?: 0 }
        .take(UiConstants.MOST_PLAYED_LIMIT)

    // Recently played
    val recentlyPlayed =
      songs
        .filter { !it.datePlayed.isNullOrBlank() }
        .sortedByDescending { it.datePlayed ?: "" }
        .take(UiConstants.RECENTLY_PLAYED_LIMIT)

    // Group songs by album for album-based sections
    val albumsMap =
      songs
        .filter { !it.albumId.isNullOrBlank() }
        .groupBy { it.albumId.orEmpty() }

    // Recently added albums - by dateCreated of first song
    val recentlyAddedAlbums =
      albumsMap
        .map { (albumId, albumSongs) ->
          val firstSong = albumSongs.maxByOrNull { it.dateCreated ?: "" } ?: albumSongs.first()
          AlbumItemWithDate(
            album =
              AlbumItem(
                id = albumId,
                name = firstSong.album ?: "Unknown Album",
                artist = firstSong.artists?.firstOrNull() ?: "Unknown Artist",
                albumArtUrl = firstSong.albumArtUrl,
                songCount = albumSongs.size,
              ),
            dateCreated = firstSong.dateCreated ?: "",
          )
        }.sortedByDescending { it.dateCreated }
        .take(UiConstants.ALBUM_SECTION_LIMIT)
        .map { it.album }

    // Random albums for "From Your Library"
    val randomAlbums =
      albumsMap
        .map { (albumId, albumSongs) ->
          val firstSong = albumSongs.first()
          AlbumItem(
            id = albumId,
            name = firstSong.album ?: "Unknown Album",
            artist = firstSong.artists?.firstOrNull() ?: "Unknown Artist",
            albumArtUrl = firstSong.albumArtUrl,
            songCount = albumSongs.size,
          )
        }.shuffled()
        .take(UiConstants.ALBUM_SECTION_LIMIT)

    // Featured albums - random selection with album art
    val featuredAlbums =
      albumsMap
        .filter { (_, albumSongs) -> albumSongs.any { !it.albumArtUrl.isNullOrBlank() } }
        .map { (albumId, albumSongs) ->
          val firstSong = albumSongs.first()
          FeaturedAlbum(
            id = albumId,
            name = firstSong.album ?: "Unknown Album",
            artist = firstSong.artists?.joinToString(", ") ?: "Unknown Artist",
            albumArtUrl = firstSong.albumArtUrl,
            songCount = albumSongs.size,
          )
        }.shuffled()
        .take(UiConstants.FEATURED_ALBUMS_LIMIT)

    mutableState.update {
      it.copy(
        isLoading = false,
        featuredAlbums = featuredAlbums,
        mostPlayed = mostPlayed,
        recentlyPlayed = recentlyPlayed,
        recentlyAddedAlbums = recentlyAddedAlbums,
        randomAlbums = randomAlbums,
      )
    }
  }

  private fun toAlbumItem(album: uniffi.aurelia_core.Album): AlbumItem =
    AlbumItem(
      id = album.id ?: "",
      name = album.name,
      artist = album.artist,
      albumArtUrl = album.albumArtUrl,
      songCount = album.songCount.toInt(),
    )

  private fun toFeaturedAlbum(album: uniffi.aurelia_core.Album): FeaturedAlbum =
    FeaturedAlbum(
      id = album.id ?: "",
      name = album.name,
      artist = album.artist,
      albumArtUrl = album.albumArtUrl,
      songCount = album.songCount.toInt(),
    )

  fun nextFeaturedAlbum() {
    val current = mutableState.value
    if (current.featuredAlbums.isNotEmpty()) {
      val nextIndex = (current.currentFeaturedIndex + 1) % current.featuredAlbums.size
      mutableState.update { it.copy(currentFeaturedIndex = nextIndex) }
    }
  }

  fun previousFeaturedAlbum() {
    val current = mutableState.value
    if (current.featuredAlbums.isNotEmpty()) {
      val prevIndex =
        if (current.currentFeaturedIndex > 0) {
          current.currentFeaturedIndex - 1
        } else {
          current.featuredAlbums.size - 1
        }
      mutableState.update { it.copy(currentFeaturedIndex = prevIndex) }
    }
  }

  fun setFeaturedIndex(index: Int) {
    val current = mutableState.value
    if (index in current.featuredAlbums.indices) {
      mutableState.update { it.copy(currentFeaturedIndex = index) }
    }
  }

  /**
   * Play a song from a specific list, setting up the queue from that list
   */
  fun playSongFromList(
    songId: String,
    songList: List<Song>,
  ) {
    val serverUrl = sessionStore.getServerUrl() ?: return
    val token = sessionStore.getToken() ?: return

    val startIndex = songList.indexOfFirst { it.id == songId }
    if (startIndex < 0) return

    mutableState.update { it.copy(currentSongId = songId) }
    playerController.setQueue(songList, serverUrl, token, startIndex)
  }

  /**
   * Play all songs from an album
   */
  fun playAlbum(albumId: String) {
    val serverUrl = sessionStore.getServerUrl() ?: return
    val token = sessionStore.getToken() ?: return

    val albumSongs =
      allSongs
        .filter { it.albumId == albumId }
        .sortedBy { it.trackNumber ?: 0 }

    if (albumSongs.isEmpty()) return

    mutableState.update { it.copy(currentSongId = albumSongs.first().id) }
    playerController.setQueue(albumSongs, serverUrl, token)
  }

  /**
   * Shuffle play an album
   */
  fun shuffleAlbum(albumId: String) {
    val serverUrl = sessionStore.getServerUrl() ?: return
    val token = sessionStore.getToken() ?: return

    val albumSongs =
      allSongs
        .filter { it.albumId == albumId }
        .shuffled()

    if (albumSongs.isEmpty()) return

    mutableState.update { it.copy(currentSongId = albumSongs.first().id) }
    playerController.setQueue(albumSongs, serverUrl, token)
  }

  fun togglePlayPause() {
    val nowPlaying = mutableState.value.nowPlaying ?: return
    if (nowPlaying.isPlaying) {
      playerController.pause()
    } else {
      playerController.resume()
    }
  }

  fun skipPrevious() {
    playerController.skipPrevious()
  }

  fun skipNext() {
    playerController.skipNext()
  }
}

private data class AlbumItemWithDate(
  val album: AlbumItem,
  val dateCreated: String,
)

private const val LOAD_FRESHNESS_MS = 60_000L
